src/main/java/org/opentripplanner/middleware/triptracker/TripInstruction.java

```diff
@@ -1,30 +1,217 @@
+            return NO_INSTRUCTION;
+        }
+        return NO_INSTRUCTION;
+    }
+
+    /**
+     * The trip instruction can only be provided if the trip is on or close to schedule.
+     */
+    private static boolean hasRequiredTripStatus(TripStatus tripStatus) {
+        return
+            !tripStatus.equals(TripStatus.NO_STATUS) &&
+            !tripStatus.equals(TripStatus.DEVIATED) &&
+            !tripStatus.equals(TripStatus.ENDED);
+    }
+
+    public static String createInstruction(Leg expectedLeg, TripSegment activeTripSegment) {
+        StepSegment activeStep = getStepEncompassingSegment(expectedLeg, activeTripSegment);
+        if (activeStep != null && isOnSimilarBearing(activeStep, activeTripSegment)) {
+            // High confidence.
+            double distanceToNextStep = Math.round(getDistance(activeStep.end, activeTripSegment.end));
+            return getInstruction(
+                distanceToNextStep,
+                getStepInstruction(expectedLeg.steps, activeStep.stepIndex + 1)
+            );
+        }
+
+        StepSegment nearestStep = getNearestStep(expectedLeg, activeTripSegment);
+
+        if (nearestStep != null) {
+            // Lower confidence. This part could benefit from "traveler bearing processing".
+            Coordinates destination = new Coordinates(expectedLeg.to);
+            if (isLastStep(nearestStep.stepIndex, expectedLeg)) {
+                if (!isTripSegmentNearerToDestination(nearestStep, activeTripSegment, destination)) {
+                    // Assuming traveler is approaching the last step.
+                    return approachingStep(activeTripSegment, nearestStep, expectedLeg);
+                }
+            } else {
+                if (isTripSegmentNearerToDestination(nearestStep, activeTripSegment, destination)) {
+                    // Assuming traveler is passed the nearest step.
+                    Step nextStep = expectedLeg.steps.get(nearestStep.stepIndex + 1);
+                    return getInstruction(
+                        getDistance(activeTripSegment.end, new Coordinates(nextStep)),
+                        nextStep.relativeDirection
+                    );
+                } else {
+                    // Assuming traveler is approaching the nearest step.
+                    return approachingStep(activeTripSegment, nearestStep, expectedLeg);
+                }
+            }
+        }
+        return TripInstruction.NO_INSTRUCTION;
+    }
+
+    /**
+     * Does the step index reference the last step in the leg.
+     */
+    private static boolean isLastStep(int stepIndex, Leg expectedLeg) {
+        return stepIndex == (expectedLeg.steps.size() - 1);
+    }
+
+    /**
+     * Provide instruction for approaching a step.
+     */
+    private static String approachingStep(TripSegment activeTripSegment, StepSegment nearestStep, Leg expectedLeg) {
+        return getInstruction(
+            getDistance(activeTripSegment.end, nearestStep.start),
```

Comment:
I am not sure about what `activeTripSegment` represents when calling the `/starttracking` endpoint. With a walk-to-transit itinerary, when calling the `/starttracking` endpoint with the start location of the first walk step, the distance computed seems to use the ending point of the `activeTripSegment` which can be away from the starting location, whereas intuitively, the distance should be with respect to the starting location of the first step. Could you clarify?

Reply:
If I have understood correctly, the preceeding method `isTripSegmentNearerToDestination` calculates distance to destination based on start coordinates, the distance calculated here should also use the (activeTripSegment) start coordinates.